enhancements/machine-config/certificate-authorities-for-image-registries.md

```diff
@@ -0,0 +1,185 @@
+essential cluster operator, and have the image regsitry operator only manage
+the optional registry workload.
+
+### User Stories
+
+* As a <role>, I want to <take some action> so that I can <accomplish a
+goal>.
+
+### Goals
+
+* Make additionalTrustedCA be managed by machine-config-operator.
+* Create a mechanism for the cluster-image-registry-operator to distribute
+  certificate authorities for the integrated image registry.
+* Remove the node-ca daemon set.
+
+### Non-Goals
+
+* Change additionalTrustedCA behavior for its users.
+
+## Proposal
+
+Create a controller inside cluster-config-operator that will replace the
+node-ca daemon set.
+
+This controller should handle and distribute to all nodes certificate
+authorities from the user-provided config map that is specified in
+`images.config.openshift.io/cluster`. It should also observe a config map in
+the `openshift-config-managed` namespace and merge its content with the
+user-provided config map. This will allow cluster-image-registry-operator to
+provide certificate authorities for the integated image registry without
+changing the user-provided config map.
+
+Once this controller is created, the node-ca daemon set should be removed.
+
+### Workflow Description
+
+Let's suppose the user creates a config map with the certificate authorities
+for their registries:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: registry-ca
+  namespace: openshift-config
+data:
+  registry.example.com: |
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+  registry-with-port.example.com..5000:
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+```
+
+Then updates the `images.config.openshift.io/cluster` object to point to this
+config map:
+
+```yaml
+spec:
+  additionalTrustedCA:
+    name: registry-ca
+```
+
+If the image registry operator is installed, there should be a config map:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: image-registry-ca
+  namespace: openshift-config-managed
+data:
+  image-registry.openshift-image-registry.svc..5000: |
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+  image-registry.openshift-image-registry.svc.cluster.local..5000: |
+    -----BEGIN CERTIFICATE-----
+    ...
+    -----END CERTIFICATE-----
+```
+
+The cluster-config-operator should update every node to propagete these
+certificate authorities to CRI-O, so it should create these files:
+
+* `/etc/docker/certs.d/registry.example.com/ca.crt`
+* `/etc/docker/certs.d/registry-with-port.example.com:5000/ca.crt`
+* `/etc/docker/certs.d/image-registry.openshift-image-registry.svc:5000/ca.crt`
+* `/etc/docker/certs.d/image-registry.openshift-image-registry.svc.cluster.local:5000/ca.crt`
+
+If the config maps are updated, old certificate authorities should be removed
+and new ones should be added. Nodes nor applications shouldn't be restarted to
+apply changes.
```

Comment:
Will these changes require [a CRI-O reload][1] to take effect?  If so, and if MachineConfig end up being used as the delivery mechanism, [this][2] will need a bump.

[1]: https://github.com/openshift/machine-config-operator/blob/3773d380e388139512b75a1278f17e0fbdd373f0/pkg/daemon/update.go#L55
[2]: https://github.com/openshift/machine-config-operator/blob/3773d380e388139512b75a1278f17e0fbdd373f0/pkg/daemon/update.go#L453-L457

Reply:
The [existing mechanism](https://github.com/openshift/cluster-image-registry-operator/blob/13471c888ed00d28989cba458b80b5b14685d89e/bindata/nodecadaemon.yaml#L44-L67) doesn't reload anything, and it seems to work.

Reply:
Trevor is right here, if we need to avoid disruption here then we'll need to add these paths to the live-apply set.

Reply:
It may actually need a restart for these fields. I can't find the special handling for a CRI-O reload for this path

Reply:
@dmage and @haircommander Did we ever reach conclusion on this? Oleg says it works currently without reloading cri-o but Peter says no.

Reply:
@mtrmac do we read the `/etc/docker/certs.d` each time we pull or is it cached at all? I would have expected it to be cached similarly to registries.conf but the current behavior imlies it's not

Reply:
That’s https://github.com/containers/image/blob/bb66acc37f166c03470cb58d3b8c808c288e1c2a/docker/docker_client.go#L268 , and it’s not cached at all - it’s read on every pull afresh (and in case of CRI-O, actually multiple times per pull). That’s not a documented/committed behavior, but I also can’t see any reason that would motivate us to change.

(And yes, `registries.conf` is cached. Sadly, the (too many) config files in c/image were added over time and are not very consistent in their file formats, lookup locations, or behaviors.)